# Cryptography (CC4017) -- Week 08


## Challenge 1

In a public-key system using RSA, you intercept the ciphertext C = 20 sent to a user whose public key is e = 13, n = 77. What is the plaintext M ?

1 < e < Φ(77)

d*e mod Φ(77) ≡  1 

Φ(77) = 60

so this becomes,

d*13 mod 60 ≡  1 

which leads to d = 37 

therefore, C^d mod (n) = M, 20³⁷ mod (77) = M = 48





## Challenge 2

In a RSA system, the public key of a given user is e = 65, n = 2881. What is the private key of this user?

Having a given e and an n we can determine the value of d, the private key, using the extended version of Euclids algorithm:

65 * x + Φ(2881) * y = 1

2881 = 43*67, Φ(2881) = 42 * 66 =  2772

65 * x + 2772 * y = 1.

In [1]:
def gcdExtended(a, b): 
    # Base Case 
    if a == 0 : 
        return b,0,1
             
    gcd,x1,y1 = gcdExtended(b%a, a) 
     
    # Update x and y using results of recursive 
    # call 
    x = y1 - (b//a) * x1 
    y = x1 
     
    return gcd,x,y 
     
 
# Driver code 
a, b = 2772, 65
g, x, y = gcdExtended(a, b) 
print(g,x,y) 

1 -17 725


We obtain x = 725, in other words, d or the private key is 725.

## Challenge 3

In the RSA public-key encryption scheme, each user has a public key e and a private key d. Suppose Bob leaks his private key. Rather than generating a new modulus, he decides to generate a new public key e and a new private key d. Is this safe?


This is not safe, having e and d an attacker can efficiently find out the values of p and q. This means that even after e and d are changed, Φ(n) can be calculated and everything else figured out, making RSA unsecure.

## Challenge 4

Suppose Bob uses the RSA cryptosystem with a very large modulus n for which the factorisation cannot be found in a reasonable amount of time. Suppose Alice sends an enciphered
message to Bob containing only her phone number: number^e (mod n). Is this safe?

While by itself there seems no problem in implementation since it seems to follow how the basic RSA implementation is done, the lack of padding and randomness makes it deterministic. In this way, if the attacker knows the secret message is a phone number, all they have to do is build a dictionary with all possible phone numbers using Bob's e and the ppublicly known n. This way they can find the value of the phone number, making it not secure.

## Challenge 5

Although, since 2002, there is a published algorithm with polynomial complexity to test
primality of an integer, its performance for small sizes is too slow to be considered as usable.
What is normally used is a probabilistic test, that can be iterated the necessary number of
times so that the probability of a false positive may be made negligible. The Miller-Rabin
is a primality test of this kind.

We can, then, write a programming function, Witness, that takes a number n and a “witness” a, with (a, n) = 1, and tests if ad ∕≡ 1 (mod n) and a2 d ∕≡ −1 (mod n), for all
0 ≤ i ≤ k. If the test succeeds we know for sure that the number is not a prime. If it fails
We cannot conclude, but we have a probability of 12 of n being a prime. We can repeat the
test (with a different values for a). If we try m times and all the tests are negative we can
ensure that the number n is a prime with a probability 1 − 2−m .
Programming assignment: Write a python program that implements this strategy and
test it for large primes.

In [35]:
import random
import math

def witness(a,n):
    
    k = 0
    d = n-1
    while (d%2 == 0):
        k += 1
        d //= 2
           
    print(d)
    print(k)
    
    if(pow(a,d,n) == 1): return False 
    
    
    for i in range (k+1):
        if(pow(a,2**i * d, n) == n-1):
            return False
    
    return True

def pick_a(n, lista):
    flag = True
    while flag:
        a = random.randint(2,n-2)
        if(math.gcd(a,n)==1 and a not in lista): flag = False
    return a
    
def prime_checker(n,t):
    if (n <= 1 or not isinstance(n, int)):
        return  False
    if n == 2:
        return True
    elif(n % 2 == 0):
        return False 
    else:
        prime_flag = True
        a_list = []
        for _ in range (t):
            a = pick_a(n, a_list)
            if(witness(a,n)):
                prime_flag = False
                break;
        return prime_flag
    


prime_checker(505,1)


    

63
3


False